benchmark-java/src/main/java/akka/grpc/benchmarks/qps/AsyncServer.java

```diff
@@ -0,0 +1,150 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package akka.grpc.benchmarks.qps;
+
+import akka.NotUsed;
+import akka.actor.ActorSystem;
+import akka.grpc.benchmarks.Utils;
+import akka.grpc.benchmarks.proto.BenchmarkService;
+import akka.grpc.benchmarks.proto.BenchmarkServiceHandlerFactory;
+import akka.grpc.benchmarks.proto.Messages;
+import akka.http.javadsl.ConnectWithHttps;
+import akka.http.javadsl.Http;
+import akka.stream.ActorMaterializer;
+import akka.stream.KillSwitches;
+import akka.stream.Materializer;
+import akka.stream.SharedKillSwitch;
+import akka.stream.javadsl.Sink;
+import akka.stream.javadsl.Source;
+import com.google.protobuf.ByteString;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.logging.Logger;
+
+/**
+ * QPS server using the non-blocking API.
+ */
+public class AsyncServer {
+  private static final Logger log = Logger.getLogger(AsyncServer.class.getName());
+
+  private BenchmarkServiceImpl benchmarkService;
+
+  /**
+   * checkstyle complains if there is no javadoc comment here.
+   */
+  public static void main(String... args) throws Exception {
+    ServerConfiguration.Builder configBuilder = ServerConfiguration.newBuilder();
+    ServerConfiguration config;
+    try {
+      config = configBuilder.build(args);
+    } catch (Exception e) {
+      System.out.println(e.getMessage());
+      configBuilder.printUsage();
+      return;
+    }
+
+    new AsyncServer().run((InetSocketAddress) config.address);
+  }
+
+  /** Equivalent of "main", but non-static. */
+  public void run(InetSocketAddress address) throws Exception {
+    // important to enable HTTP/2 in ActorSystem's config
+    Config conf = ConfigFactory.parseString("akka.http.server.preview.enable-http2 = on")
+        .withFallback(ConfigFactory.defaultApplication());
+    ActorSystem system = ActorSystem.create("LoadWorker", conf);
+
+    Materializer mat = ActorMaterializer.create(system);
+
+    benchmarkService = new BenchmarkServiceImpl(mat);
+
+    Http.get(system).bindAndHandleAsync(
+        BenchmarkServiceHandlerFactory.create(benchmarkService, mat),
+        ConnectWithHttps.toHostHttps(address.getHostName(), address.getPort()).withCustomHttpsContext(Utils.serverHttpContext()),
+        mat)
+        .thenAccept(binding -> {
+          System.out.println("gRPC server bound to: " + address);
+        });
+  }
+
+  public void shutdown() {
+    if (benchmarkService != null)
+      benchmarkService.shutdown();
+    benchmarkService = null;
+  }
+
+  public static class BenchmarkServiceImpl implements BenchmarkService {
```

Comment:
It's rather pleasing to compare the implementation of this service with https://github.com/grpc/grpc-java/blob/master/benchmarks/src/main/java/io/grpc/benchmarks/qps/AsyncServer.java#L194

Reply:
Indeed, much much nicer :-)